src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala

```diff
@@ -0,0 +1,92 @@
+package fr.polytechnique.cmap.cnam.etl.transformers.follow_up
```

Comment:
As this class has many problems, I would suggest the following implementation:
```scala
package fr.polytechnique.cmap.cnam.etl.transformers.follow_up

import java.sql.Timestamp
import org.joda.time.{DateTime, DateTimeZone}
import org.apache.spark.sql.Dataset
import fr.polytechnique.cmap.cnam.etl.events.{Event, FollowUp}
import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.Columns.EndReasons

object FollowUpTransformerUtilities {


  case class PatientDates(
    patientID: String,
    deathDate: Option[Timestamp],
    followUpStart: Option[Timestamp],
    observationEnd: Option[Timestamp])


  case class TrackLossDate(
    patientID: String,
    trackloss: Option[Timestamp])

  sealed trait FollowUpEndReason extends Timestamp{
    val reason: String
  }

  class Death extends FollowUpEndReason {
    val reason = EndReasons.Death.toString
  }

  class Disease extends FollowUpEndReason {
    override val reason: String = EndReasons.Disease.toString
  }

  class TrackLoss extends FollowUpEndReason {
    override val reason: String = EndReasons.Trackloss.toString
  }

  class ObservationEnd extends FollowUpEndReason {
    override val reason: String = EndReasons.ObservationEnd.toString
  }


  val correctedStart: (Timestamp, Option[Timestamp], Int) => Option[Timestamp] =
    (start: Timestamp, end: Option[Timestamp], delayMonths: Int) => {
      val st = Timestamp.valueOf(
        new DateTime(start, DateTimeZone.UTC).plusMonths(delayMonths)
          .toString("YYYY-MM-dd HH:mm:ss")
      )
      val corrected: Option[Timestamp] = if (st.before(end.get)) Some(st) else None
      corrected
    }

  val tracklossDateCorrected: (Timestamp, Timestamp) => Option[Timestamp] =
    (start: Timestamp, followUpStart: Timestamp) => {
      val corrected: Option[Timestamp] = if (start.after(followUpStart)) Some(start) else None
      corrected
    }
  

  val endReason: (Option[Timestamp], Option[Timestamp], Option[Timestamp], Option[Timestamp]) => Option[FollowUpEndReason] =
    (
      observationEnd: Option[Timestamp],
      deathDate: Option[Timestamp],
      trackloss: Option[Timestamp], 
      firstTargetDiseaseDate: Option[Timestamp]) => {

      Seq(
        trackloss.map(_.asInstanceOf[TrackLoss]),
        deathDate.map(_.asInstanceOf[Death]),
        observationEnd.map(_.asInstanceOf[ObservationEnd]),
        firstTargetDiseaseDate.map(_.asInstanceOf[Disease])
      ).min
    }


  implicit class FollowUpDataset(followups: Dataset[Event[FollowUp]]) {
    def cleanFollowUps(): Dataset[Event[FollowUp]] = followups.filter(e => e.start.before(e.end.get))
  }

}
```

Reply:
I agree with you.

Reply:
I don't mind to change it, but this implementation throw me lot of errors

Reply:
What are the thrown errors?

Reply:
For instance:

Error:(25, 23) overloaded method constructor Timestamp with alternatives:
  (x$1: Long)java.sql.Timestamp <and>
  (x$1: Int,x$2: Int,x$3: Int,x$4: Int,x$5: Int,x$6: Int,x$7: Int)java.sql.Timestamp
 cannot be applied to ()
  class Death extends FollowUpEndReason {

Reply:
I'm trying to implement it and I found this error as well Error:(72, 9) No implicit Ordering defined for Option[fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformerUtilities.FollowUpEndReason].
      ).min 

I can't use implicit ordering with Option[FollowUpEndReason]

Reply:
For the first, just add the date as an argument and define the case class.

As for the second, define an implicit value in the scope that defines the Ordering for the type. I taught that inheriting from Timestamp would have defined the Ordering, which is not the case.